File: Version-1/C++/MultipleChoiceQuestion.cpp
#include "MultipleChoiceQuestion.h"
#include <iostream>
#include <algorithm>

MultipleChoiceQuestion::MultipleChoiceQuestion(
    const std::string& query,
    const std::string& a,
    const std::string& b,
    const std::string& c,
    const std::string& d,
    const std::string& e,
    const std::string& answer
) {
    question = query + "\n";
    question += "A. " + a + "\n";
    question += "B. " + b + "\n";
    question += "C. " + c + "\n";
    question += "D. " + d + "\n";
    question += "E. " + e + "\n";

    correctAnswer = answer;
    std::transform(correctAnswer.begin(),
		   correctAnswer.end(),
  		   correctAnswer.begin(),
		   ::toupper);
}

std::string MultipleChoiceQuestion::ask() {
    while (true) {
	std::cout << question;
	std::cout << "Your anser: ";

	std::string myAnswer;
	std::getline(std::cin, myAnswer);

	std::transform(myAnswer.begin(),
		       myAnswer.end(),
		       myAnswer.begin(),
		       ::toupper);

	if (myAnswer == "A" || myAnswer == "B" ||
	    myAnswer == "C" || myAnswer == "D" ||
	    myAnswer == "E") {
	    return myAnswer;
	} else {
	    std::cout << "Invalid answer. Please enter A, B, C, D or E.\n";
	}
}
 

File: Version-1/Java/MultipleChoiceQuestion.java
import javax.swing.JOptionPane;

public class MultipleChoiceQuestion extends Question {

    // static int nQuestion = 0;
    // static int nCorrect = 0;

    // String question;
    // String correctAnswer;

    public MultipleChoiceQuestion(
        String query, 
        String a, 
        String b, 
        String c, 
        String d, 
        String e, 
        String answer
    ) { 
        question = query + "\n";
        question += "A. " + a + "\n";	
        question += "B. " + b + "\n";	
        question += "C. " + c + "\n";	
        question += "D. " + d + "\n";	
        question += "E. " + e + "\n";	

        correctAnswer = answer;
        correctAnswer = correctAnswer.toUpperCase();
    }

    String ask() {

        while (true) {
	    String myAnswer = JOptionPane.showInputDialog(question);
	    myAnswer = myAnswer.toUpperCase();

           if (myAnswer.equals("A") || myAnswer.equals("B") ||
               myAnswer.equals("C") || myAnswer.equals("D") || myAnswer.equals("E")) {
               return myAnswer;
           } else {
               JOptionPane.showMessageDialog(null, "Invalid answer." 
                                             " Please enter A, B, C, D or E.");
               continue;
           }
        }	    
    }
}    
